### PR TITLE
KAFKA-4031: Check if buffer cleaner is null before using it

### PR DESCRIPTION
**Pull request by soumyajit-sahu**
A small fix to check null before using the reference

### CODE REVIEW

File: core/src/main/scala/kafka/log/OffsetIndex.scala [hunk 1]
```diff
@@ -309,7 +309,12 @@ class OffsetIndex(@volatile private[this] var _file: File, val baseOffset: Long,
   private def forceUnmap(m: MappedByteBuffer) {
     try {
       m match {
-        case buffer: DirectBuffer => buffer.cleaner().clean()
+        case buffer: DirectBuffer => {
+          val bufferCleaner = buffer.cleaner()
+          if (bufferCleaner != null) {
+            bufferCleaner.clean()
+          }
+        }
```
Comment (ijuma) [inline core/src/main/scala/kafka/log/OffsetIndex.scala]: Nitpick: too many braces. :)

``` scala
case buffer: DirectBuffer =>
  val bufferCleaner = buffer.cleaner()
  if (bufferCleaner != null)
    bufferCleaner.clean()
```

Comment (soumyajit-sahu) [inline core/src/main/scala/kafka/log/OffsetIndex.scala]: Thanks! Removed them

**Comment by ijuma** [comment]
Do you know under what scenarios `cleaner` can be null?

**Comment by soumyajit-sahu** [comment]
It has eluded me, but after reading code from Kafka through OpenJDK's, this is my guess.
If the following call stack happened:
OffsetIndex._entries = 0 //the "0" here triggers the condition for the cleaner to be null
OffsetIndex.truncateToEntries(0)
OffsetIndex.truncate()
LogSegment.recover()

Followed by another call stack say from LogCleaner sometime:
raf.getChannel().map(FileChannel.MapMode.READ_WRITE, 0, roundedNewSize) //roundedNewSize=0
OffsetIndex.resize(_entries \* 8) //basically resize(0)
OffsetIndex.trimToValidSize()

Now, when roundedNewSize = 0, the OpenJDK's sun.nio.ch.FileChannelImpl.map() will return "Util.newMappedByteBufferR(0, 0, dummy, null)". Notice that the last param is null. That sets the cleaner inside the map to be null too.

**Comment by ijuma** [comment]
Thanks for the investigation. Makes sense to me. Can you please add a brief comment in the code? Something like:

"`cleaner` can be null if the mapped region has size 0"

I will then merge it.

**Comment by soumyajit-sahu** [comment]
Added a comment. Thanks!

**Comment by ijuma** [comment]
LGTM

**Pull request closed without merging**